### DISCUSSION TITLE
Local AI backends for OPUS-MT and other HuggingFace models.

### DISCUSSION THREAD

**Discussion by Nigel2392** [wagtail/wagtail-localize#764] [Ideas]
I'm wondering; is anyone interested / would it be possible in this project to integrate local LLM backends? (Of course I will publish my code with tests)
These might be useful for developers wanting to translate locally; without cost. (There are some really low resource models out there!)
Allows for more leniency - but the drawback, most translation models I could find which are reasonably performant do not support HTML translation.

I currently in my own project have the following block which only allows for the local AI backend if `transformers` is installed - wondering if this would be the recommended way of implementing it.
```python
try:
   from transformers import (...)

   class OpusTranslator(...):
      ...
except ImportError:
   OpusTranslator = None
```

**Comment by chris48s** [comment]
wagtail-localize has support for several machine translators already:

- deepl
- google cloud translate
- libre translate

each of these is an implementation of `BaseMachineTranslator`

All 3 of these implement the `translate()` method by sending text off to an external API and returning the result, but there's no reason why the `translate()` method couldn't do the translation using a locally installed LLM.

I'd suggest having a look over https://github.com/wagtail/wagtail-localize/tree/bb04c7627701c7922be66389a1aaa2a721a09559/wagtail_localize/machine_translators

Sounds like what you want to do could be achieved by making an implementation of `BaseMachineTranslator`, which can then be used in `WAGTAILLOCALIZE_MACHINE_TRANSLATOR["CLASS"]`.

**Reply by Nigel2392** [reply]
Yes! I do already have this - the OpusTranslator is already a subclass of BaseMachineTranslator. Just wondering if anyone would be interested. :)

**Reply by zerolab** [reply]
Build it and they will come ;)
I am quite sure if you package it and share in the Wagtail Slack, folks using wagtail-localize will at the very least play with it

**Comment by zerolab** [comment]
To add to what @chris48s shared, I'd suggest making that its own module/package as you can iterate faster when it is on its own.
